samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala

```diff
@@ -161,72 +127,102 @@ class KafkaSystemProducer(systemName: String,
+        throw producerException
     }
   }
 
+
   def flush(source: String) {
     updateTimer(metrics.flushNs) {
       metrics.flushes.inc
 
-      val sourceData = sources.get(source)
-      //if latestFuture is null, it probably means that there has been no calls to "send" messages
-      //Hence, nothing to do in flush
-      if(sourceData.latestFuture != null) {
-        while(!sourceData.latestFuture.isDone && sourceData.exceptionInCallback.get() == null) {
-          try {
-            sourceData.latestFuture.get()
-          } catch {
-            case t: Throwable => error(t.getMessage, t)
-          }
-        }
+      if (producer == null) {
+        throw new SystemProducerException(KafkaSystemProducer.NullProducerMsg)
+      }
 
-        //if there is an exception thrown from the previous callbacks just before flush, we have to fail the container
-        if (sourceData.exceptionInCallback.get() != null) {
-          metrics.flushFailed.inc
-          throw sourceData.exceptionInCallback.get()
-        } else {
-          trace("Flushed %s." format (source))
+      // Flush only throws InterruptedException, all other errors are handled in send() callbacks
+      producer.flush()
+
+      // Invariant: At this point either
+      // 1. The producer is fine and there are no exceptions to handle   OR
+      // 2. The producer is closed and one or more sources have exceptions to handle
+      //   2a. All new sends get a ProducerClosedException or IllegalStateException (depending on kafka version)
+      //   2b. There are no messages in flight because the producer is closed
+
+      // We must check for an exception AFTER flush() because when flush() returns all callbacks for messages sent
+      // in that flush() are guaranteed to have completed and we update the exception in the callback.
+      // If there is an exception, we rethrow it here to prevent the checkpoint.
+      val exception = fatalException.get()
+      if (exception != null) {
+        metrics.flushFailed.inc
+        throw new SystemProducerException(KafkaSystemProducer.FailedFlushMsg, exception)
+      }
+      trace("Flushed %s." format source)
+    }
+  }
+
+
+  private def handleSendException(currentProducer: Producer[Array[Byte], Array[Byte]], producerException: SystemProducerException, isFatalException: Boolean) = {
+    metrics.sendFailed.inc
+    error(producerException)
+    // The SystemProducer contract is synchronous, so there's no way for us to guarantee that an exception will
+    // be handled by the Task before we recreate the producer, and if it isn't handled, a concurrent send() from another
+    // source could send on the new producer before handling its own exceptions and produce out of order messages.
+    // So we have to handle it right here in the SystemProducer.
+    if (dropProducerExceptions) {
+      warn(KafkaSystemProducer.DropErrorMsg, producerException)
```

Comment:
This can potentially create a large amount of logs when the Kafka cluster being written to is unavailable.
We should consider adding a throttling logger for messages like these, but for now we can remove the producerException from the message (it's already logged as error above).